modules/queue/manager.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package queue
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+	"sync"
+	"time"
+
+	"code.gitea.io/gitea/modules/log"
+)
+
+var manager *Manager
+
+// Manager is a queue manager
+type Manager struct {
+	mutex sync.Mutex
+
+	counter int64
+	Queues  map[int64]*Description
+}
+
+// Description represents a working queue inheriting from Gitea.
+type Description struct {
+	mutex           sync.Mutex
+	QID             int64
+	Queue           Queue
+	Type            Type
+	Name            string
+	Configuration   interface{}
+	ExemplarType    string
+	addWorkers      func(number int, timeout time.Duration) context.CancelFunc
+	numberOfWorkers func() int
+	counter         int64
+	PoolWorkers     map[int64]*PoolWorkers
+}
+
+// DescriptionList implements the sort.Interface
+type DescriptionList []*Description
+
+// PoolWorkers represents a working queue inheriting from Gitea.
+type PoolWorkers struct {
+	PID        int64
+	Workers    int
+	Start      time.Time
+	Timeout    time.Time
+	HasTimeout bool
+	Cancel     context.CancelFunc
+}
+
+// PoolWorkersList implements the sort.Interface
+type PoolWorkersList []*PoolWorkers
+
+func init() {
+	_ = GetManager()
+}
+
+// GetManager returns a Manager and initializes one as singleton if there's none yet
+func GetManager() *Manager {
+	if manager == nil {
```

Comment:
`once.Do()`? Otherwise a race condition might occur.

Reply:
It's created at init so there's no race.